src/content/reference/react-dom/components/form.md

```diff
@@ -0,0 +1,433 @@
+
+<Sandpack>
+
+
+```js App.js
+import { useOptimistic, useState, useRef } from "react";
+import { deliverMessage } from "./actions.js";
+
+function Thread({ messages, sendMessage }) {
+  const formRef = useRef();
+  async function formAction(formData) {
+    addOptimisticMessage(formData.get("message"));
+    formRef.current.reset();
+    await sendMessage(formData);
+  }
+  const [optimisticMessages, addOptimisticMessage] = useOptimistic(
+    messages,
+    (state, newMessage) => [
+      ...state,
+      {
+        text: newMessage,
+        sending: true
+      }
+    ]
+  );
+
+  return (
+    <>
+      {optimisticMessages.map((message, index) => (
+        <div key={index}>
+          {message.text}
+          {!!message.sending && <small> (Sending...)</small>}
+        </div>
+      ))}
+      <form action={formAction} ref={formRef}>
+        <input type="text" name="message" placeholder="Hello!" />
+        <button type="submit">Send</button>
+      </form>
+    </>
+  );
+}
+
+export default function App() {
+  const [messages, setMessages] = useState([
+    { text: "Hello there!", sending: false, key: 1 }
+  ]);
+  async function sendMessage(formData) {
+    const sentMessage = await deliverMessage(formData.get("message"));
+    setMessages([...messages, { text: sentMessage }]);
+  }
+  return <Thread messages={messages} sendMessage={sendMessage} />;
+}
+```
+
+```js actions.js
+export async function deliverMessage(message) {
+  await new Promise((res) => setTimeout(res, 1000));
+  return message;
+}
+```
+
+
+```json package.json hidden
+{
+  "dependencies": {
+    "react": "18.3.0-canary-6db7f4209-20231021",
+    "react-dom": "18.3.0-canary-6db7f4209-20231021",
+    "react-scripts": "^5.0.0"
+  },
+  "main": "/index.js",
+  "devDependencies": {}
+}
+```
+
+</Sandpack>
+
+[//]: # 'Uncomment the next line, and delete this line after the `useOptimisitc` reference documentatino page is published'
+[//]: # 'To learn more about the `useOptimistic` Hook see the [reference documentation](/reference/react/hooks/useOptimistic).'
+
+### Handling form submission errors {/*handling-form-submission-errors*/}
+
+In some cases the function called by a `<form>`'s `action` prop throw an error. You can handle these errors by wrapping `<form>` in an Error Boundary. If the function called by a `<form>`'s `action` prop throws an error, the fallback for the error boundary will be displayed.
+
+<Sandpack>
+
+```js App.js
+import { ErrorBoundary } from "react-error-boundary";
+import { action } from "./actions.js"
+
+export default function Search() {
+  return (
+    <ErrorBoundary fallback={<p>There was an error while submitting the form</p>}>
```

Comment:
Not sure why but the fallback error message isn't displaying for me. Is this expected?

https://github.com/reactjs/react.dev/assets/20743223/a2726e61-c7b6-49ad-aa03-3b317c5943ca

Reply:
same, I'm wondering if it's because its not using RSCs? Or do we maybe need to wrap this in a startTransition?

Reply:
Does it work if you have the contents in a different component?

Reply:
@TheSavior should be equivalent since there aren't hooks in Search. ErrorBoundary doesn't care if it goes `ErrorBoundary > form` or `ErrorBoundary > Foo > form`

Reply:
Moving the client action into the component seems to have fixed the issue. I'll look into why this version of the code didn't work later.